### PR TITLE
Optimize GroupBy's aggregation algorithm

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Now we decide aggregation strategy at runtime described in #2215 , this PR's optimization try to avoid some bad choices for tile method.

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/dataframe/groupby/aggregation.py [hunk 1]
```diff
@@ -659,15 +660,26 @@ def _tile_auto(
...
+        # almost equals to the tileable's, then use tree method
+        # as combine aggregation results won't lead to a rapid expansion.
+        ratios = [
+            agg_size / raw_size for agg_size, raw_size in zip(agg_sizes, raw_sizes)
+        ]
+        cv = variation(agg_sizes)
+        mean_ratio = np.mean(ratios)
+        if cv <= 0.2 and mean_ratio <= 0.8:
+            return cls._combine_tree(op, chunks + left_chunks, out_df, func_infos)
+        elif mean_ratio <= 0.25:
```
Comment (qinxuye) [inline mars/dataframe/groupby/aggregation.py:681]: len(chunks) is `combine_size`, should not be specified as 1/4 directly.

File: mars/dataframe/groupby/tests/test_groupby_execution.py [hunk 2]
```diff
@@ -580,12 +580,18 @@ def _disallow_reduce(ctx, op):
...
+            "c1": list(range(4)) * 5,
+            "c2": rs.choice(["a", "b", "c"], (20,)),
+            "c3": rs.rand(20),
+        }
+    )
 
+    mdf = md.DataFrame(raw, chunk_size=5)
     r = mdf.groupby("c1").agg("sum")
-    operand_executors = {DataFrameGroupByAgg: _disallow_combine_and_agg}
+    operand_executors = {DataFrameGroupByAgg: _disallow_reduce}
```
Comment (qinxuye) [inline mars/dataframe/groupby/tests/test_groupby_execution.py:594]: Why remove the original case that may trigger shuffle groupby?

Comment (hekaisheng) [reply mars/dataframe/groupby/tests/test_groupby_execution.py:605]: For current strategy, the original case will choose tree method with high probability, I add a new case that will choose shuffle for sure.

**Comment by hekaisheng** [comment]
> The patch coverage seems not passed, please check the reason.

There indeed is a certain probability that one branch doesn't be covered, changed now.

**Comment by qinxuye** [comment]
Coverage goes down again, is it relevant to the modification that changes threshold from 0.8 to 2/3?

**Review by qinxuye — APPROVED** [review]
LGTM

**Review by wjsi — APPROVED** [review]
LGTM

**Pull request merged**